testProjectKotlinDslBase/build.gradle.kts

```diff
@@ -0,0 +1,174 @@
+tasks {
+
+    "jar"(Jar::class) {
+        sourceSets.forEach { sourceSet ->
+            from(sourceSet.output)
+
+            val compileTaskName = sourceSet.getCompileTaskName("java")
+            dependsOn(tasks.getByName(compileTaskName))
+        }
+    }
+
+    "test"{
+
+        doLast{
+            val generateProtoTasks = project.protobuf.protobuf.generateProtoTasks
+
+            val generateProtoTaskNames = generateProtoTasks.all().map { it.name }.toSet()
+            val generateProtoTaskNamesMain = generateProtoTasks.ofSourceSet("main").map { it.name }.toSet()
+
+            assert(setOf("generateProto", "generateGrpcProto", "generateTestProto") == generateProtoTaskNames)
+            assert(setOf("generateProto") == generateProtoTaskNamesMain)
+
+            assertJavaCompileHasProtoGeneratedDir("main", listOf("java"))
+            assertJavaCompileHasProtoGeneratedDir("test", listOf("java"))
+            assertJavaCompileHasProtoGeneratedDir("grpc", listOf("java", "grpc_output"))
+
+            listOf("main", "test").forEach { sourceSet ->
+                assertFileExists(false, "$buildDir/generated/source/proto/$sourceSet/descriptor_set.desc")
+            }
+            assertFileExists(true, "$buildDir/generated/source/proto/grpc/descriptor_set.desc")
+        }
+    }
+}
+
+fun assertJavaCompileHasProtoGeneratedDir(sourceSet: String, codegenPlugins: Collection<String>) {
+    val compileJavaTask = tasks.getByName(sourceSets.getByName(sourceSet).getCompileTaskName("java")) as JavaCompile
+    assertJavaCompileHasProtoGeneratedDir(project, sourceSet, compileJavaTask, codegenPlugins)
+}
+
+fun assertFileExists(exists: Boolean, path: String) {
+    if (exists) {
+        assert(File(path).exists())
+    } else {
+        assert(!File(path).exists())
+    }
+}
+
+fun assertJavaCompileHasProtoGeneratedDir(
+    project: Project,
+    sourceSet: String,
+    compileJavaTask: JavaCompile,
+    codegenPlugins: Collection<String>
+) {
+    val baseDir = File("${project.buildDir}/generated/source/proto/$sourceSet")
+    // The expected direct subdirectories under baseDir
+    val expectedDirs = codegenPlugins.map { codegenPlugin ->
+        File("${project.buildDir}/generated/source/proto/$sourceSet/$codegenPlugin")
+    }.toSet()
+
+    val actualDirs = mutableSetOf<File>()
+    compileJavaTask.source.visit {
+
+        // If the visited file is or is under a direct subdirectory of baseDir, add
+        // that subdirectory to actualDirs.
+        var file = this@visit.file
+        while (true) {
+            if (file.parentFile == baseDir) {
+                actualDirs.add(file)
+            }
+            if (file.parentFile == null) {
+                break
+            }
+            file = file.parentFile
+        }
+    }
+    assert(expectedDirs == actualDirs)
+}
+
+// This extension is not auto generated when we apply the plugin using
+// apply(plugin = "com.google.protobuf")
+val Project.protobuf: ProtobufConvention get() =
```

Comment:
Since this is another extension that relies on the plugin being applied in the plugins block, should we add it to ```ProtobufConfiguratorExts.kt``` as well?

Reply:
This extension is used to get the plugin convention from the project, which was needed for Line 106. I didnt realize that ```HasConvention``` isnt part of the public api. The code itself I copied from the dynamically generated code file that gradle creates when you use the plugins block. I can see if there's a way around casting to ```HasConvention```

Reply:
Cleaned up the usage to the following
```
val Project.protobuf: ProtobufConvention get() =
    this.convention.getPlugin(ProtobufConvention::class)
```

Reply:
nit: Can this workaround be moved nearer to `apply(plugin = "com.google.protobuf")` ?